### PR TITLE
[fix]: fix tanh grad grad, add the third derivative test

### PR DESCRIPTION
**Pull request by Yang-lin-09**
修复之前更新 tanh op 的高阶微分规则。使用tanh_grad来计算tanh_grad_grad中grad_require_grad的部分。添加了三阶微分的测试，二阶微分的测试不会涉及到grad_require_grad这一部分。

在OneScience中的cylinder2d例子的性能变化：
origin: 0.29/epoch
now: 0.26/epoch

### CODE REVIEW

**Comment by Yang-lin-09** [comment]
oneflow的square算子三阶导会跑出来nan，torch的square算子三阶导全是0

File: python/oneflow/test/modules/test_math_op_higher_derivative.py [hunk 1]
```diff
@@ -64,6 +64,62 @@ def _test_math_op_grad_grad_impl(test_case, op_name):
     )
 
 
+def _test_tanh_grad_grad_grad_impl(test_case, op_name):
```
Comment (levi131) [inline python/oneflow/test/modules/test_math_op_higher_derivative.py:67]: _test_math_grad_grad_grad_impl

File: python/oneflow/test/modules/test_math_op_higher_derivative.py [hunk 2]
```diff
@@ -81,7 +137,7 @@ def test_cosh_grad_grad(test_case):
         _test_math_op_grad_grad_impl(test_case, "cosh")
 
     def test_tanh_grad_grad(test_case):
```
Comment (levi131) [inline python/oneflow/test/modules/test_math_op_higher_derivative.py:139]: test_tanh_grad_grad_grad

File: python/oneflow/test/modules/test_global_math_op_higher_derivative.py [hunk 3]
```diff
@@ -100,7 +157,7 @@ def test_global_cosh_grad_grad(test_case):
     def test_global_tanh_grad_grad(test_case):
```
Comment (levi131) [inline python/oneflow/test/modules/test_global_math_op_higher_derivative.py:157]: test_global_tanh_grad_grad_grad

**Comment by Yang-lin-09** [comment]
abs和square的bug放在了这个issue中：https://github.com/Oneflow-Inc/oneflow/issues/10243

**Review by levi131 — APPROVED** [review]
LGTM

File: oneflow/core/autograd/higher_order_gradient_funcs/math_unary_op.cpp [hunk 4]
```diff
@@ -102,16 +102,7 @@ class TanhGradGradCls : public OpExprGradFunction<UnaryMathGradGradState> {
...
-                                .then([](const std::shared_ptr<one::Tensor>& input) {
-                                  return functional::ScalarSub(Scalar(1), input, /* alpha */ 1);
-                                })
-                                .then([out_grad](const std::shared_ptr<one::Tensor>& input) {
-                                  return functional::Mul(input, out_grad);
-                                })
-                                .call(input));
-    }
+    if (ctx->grad_requires_grad) { (*in_grads)[1] = JUST(functional::TanhGrad(input, out_grad)); }
     return Maybe<void>::Ok();
```
Comment (reygu) [inline oneflow/core/autograd/higher_order_gradient_funcs/math_unary_op.cpp:106]: TanhGradGradCls 应该可以复用 UnaryMathGradGrad 的代码吧，一阶反向和 sigmoid 一样是接收 (y, dy) 参数，sigmoid 定义在 `MATH_UNARY_ELEMENTWISE_GRAD_GRAD_DY_Y_FUNC_SEQ`

File: python/oneflow/test/modules/test_global_math_op_higher_derivative.py [hunk 5]
```diff
@@ -97,10 +154,10 @@ def test_global_cosh_grad_grad(test_case):
                 _global_math_op_grad_grad_impl(test_case, "cosh", placement, sbp)
 
     @globaltest
-    def test_global_tanh_grad_grad(test_case):
+    def test_global_tanh_grad_grad_grad(test_case):
```
Comment (hjchen2) [inline python/oneflow/test/modules/test_global_math_op_higher_derivative.py:157]: 这里二阶导不需要测了么？

File: oneflow/core/autograd/higher_order_gradient_funcs/math_unary_op.cpp (hunk 4 above)
Comment (Yang-lin-09) [reply oneflow/core/autograd/higher_order_gradient_funcs/math_unary_op.cpp:106]: 应该是可以的，我试一下。

File: python/oneflow/test/modules/test_global_math_op_higher_derivative.py (hunk 5 above)
Comment (Yang-lin-09) [reply python/oneflow/test/modules/test_global_math_op_higher_derivative.py:157]: 这个三阶导的测试里包含了二阶导

Comment (hjchen2) [reply python/oneflow/test/modules/test_global_math_op_higher_derivative.py:157]: 要不_global_math_grad_grad_grad_impl这个接口改一下吧，改成_global_math_nth_derivative_impl(3，...)，支持求任意阶导数

Comment (hjchen2) [reply python/oneflow/test/modules/test_global_math_op_higher_derivative.py:157]: 然后把_global_math_op_grad_grad_impl删了

Comment (Yang-lin-09) [reply python/oneflow/test/modules/test_global_math_op_higher_derivative.py:157]: 这样也可以，不过现在加了这个_global_math_grad_grad_grad_impl接口是因为上面提到的两个bug，如果没那两个bug直接就可以合并成一个。

Comment (hjchen2) [reply python/oneflow/test/modules/test_global_math_op_higher_derivative.py:157]: 没关系吧，其他Op还是测2阶的

Comment (Yang-lin-09) [reply python/oneflow/test/modules/test_global_math_op_higher_derivative.py:157]: 刚刚发现了这里加这个似乎没什么必要了，因为二阶导中init_grad这个变量 requires_grad_()已经覆盖到了这个分支，这个三阶导以及接口改动的问题，可能可以和那个issue放在一起解决。

**Comment by hjchen2** [comment]
> 在两个文件的三阶导测试中，oneflow的square算子三阶导会跑出来nan，torch的square算子三阶导全是0

torch的square算子三阶导我这里是None，你用的pytorch版本是多少

**Review by reygu — APPROVED** [review]
